### PR TITLE
feat: `map_chainspec` for `NodeConfig`

### LINKED ISSUE #12040: feat: introduce a method on `NodeConfig` for replacing the generic chainspec to dynamic `EthChainSpec`.
### Describe the feature

Since `EthChainSpec` is an object-safe trait now, implement mapping from `NodeConfig<ChainSpec>` to `NodeConfig<Box<dyn EthChainSpec>>` as a function on `NodeConfig` :

https://github.com/paradigmxyz/reth/blob/bea6e278dfdb052ce7bbcec5d42349221aa04b84/crates/node/core/src/node_config.rs#L141

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by greged93**
Introduce the `erased` method on `NodeConfig`, which allows to map a `NodeConfig<ChainSpec> where ChainSpec: EthChainSpec` into a `NodeConfig<Box<dyn EthChainSpec>>`.

Resolves #12040

### CODE REVIEW

File: crates/node/core/src/node_config.rs [hunk 1]
```diff
@@ -422,6 +422,29 @@ impl<ChainSpec> NodeConfig<ChainSpec> {
             Err(e) => Err(eyre!("Failed to load configuration: {e}")),
         }
     }
+
+    /// Type erases the [`ChainSpec`] generic into a [`Box<dyn EthChainSpec>`].
+    pub fn erased(self) -> NodeConfig<Box<dyn EthChainSpec>>
```
Comment (greged93) [inline crates/node/core/src/node_config.rs:427]: could be `#[inline]`?

Comment (mattsse) [inline crates/node/core/src/node_config.rs:427]: what we want here is `fn map_chainspec<F, C> where F:FnOnce(Self) -> NodeConfig<C>`

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/node/core/src/node_config.rs (hunk 1 above)
Comment (greged93) [reply crates/node/core/src/node_config.rs:427]: ahhh got it sorry I misunderstood what you wanted, thought you just wanted to extract the erasing part. Updating.

File: crates/node/core/src/node_config.rs [hunk 2]
```diff
@@ -422,6 +422,14 @@ impl<ChainSpec> NodeConfig<ChainSpec> {
             Err(e) => Err(eyre!("Failed to load configuration: {e}")),
         }
     }
+
+    /// Maps the config using the provided closure to update the [`ChainSpec`] generic.
+    pub fn map_chainspec<F, C>(self, f: F) -> NodeConfig<C>
+    where
+        F: FnOnce(Self) -> NodeConfig<C>,
+    {
+        f(self)
```
Comment (mattsse) [inline crates/node/core/src/node_config.rs:431]: ah okay, sorry I gave somewhat conflicting suggestions here, imo we can limit this to chainspec only so the closure should actually return the chainspec type given the current chainspec as input so probably `FnOnce(ChainSpec) -> C`

then we can move all of this in this function
```
NodeConfig {
                chain,
                datadir: config.datadir,
                config: config.config,
                metrics: config.metrics,
                instance: config.instance,
                network: config.network,
                rpc: config.rpc,
                txpool: config.txpool,
                builder: config.builder,
                debug: config.debug,
                db: config.db,
                dev: config.dev,
                pruning: config.pruning,
            }
```

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/node/core/src/node_config.rs (hunk 2 above)
Comment (greged93) [reply crates/node/core/src/node_config.rs:431]: ah yeah I wasn't sure based on the pseudo code but this makes more sense haha

**Comment by greged93** [comment]
sorry about the back and forth, let me know if this is better

**Review by mattsse — APPROVED** [review]
lgtm!

ty

**Pull request merged**